### DISCUSSION TITLE
Dataloader batch - RuntimeError: Could not infer dtype of NoneType

### DISCUSSION THREAD

**Discussion by piyush-J** [pyg-team/pytorch_geometric#4554] [Q&A]
Hi! Thank you for an amazing repository. I am a student working on a problem that has multiple graphs in the dataset and one of the Data object looks something like this:

`Data(y=1, input_ids_0=[6, 128], attention_mask_0=[6, 128], edge_index_0=[2, 6], input_ids_1=[9, 128], attention_mask_1=[9, 128], edge_index_1=[2, 11], input_ids_2=[6, 128], attention_mask_2=[6, 128], edge_index_2=[2, 6], input_ids_3=[9, 128], attention_mask_3=[9, 128], edge_index_3=[2, 11], input_ids_4=[1, 128], attention_mask_4=[1, 128], edge_index_4=[2, 0], input_ids_5=[1, 128], attention_mask_5=[1, 128], edge_index_5=[2, 0], input_ids_6=[1, 128], attention_mask_6=[1, 128], edge_index_6=[2, 0], input_ids_7=[1, 128], attention_mask_7=[1, 128], edge_index_7=[2, 0])`

**CODE:**
```
from torch_geometric.loader import DataLoader
# all_data_test is a list of Data objects
test_loader = DataLoader(all_data_test, batch_size=64, shuffle=False)

for batch in test_loader:
    print(batch)
```

**ERROR:**
```
[/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py](https://localhost:8080/#) in __next__(self)
    528             if self._sampler_iter is None:
    529                 self._reset()
--> 530             data = self._next_data()
    531             self._num_yielded += 1
    532             if self._dataset_kind == _DatasetKind.Iterable and \

[/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py](https://localhost:8080/#) in _next_data(self)
    568     def _next_data(self):
    569         index = self._next_index()  # may raise StopIteration
--> 570         data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    571         if self._pin_memory:
    572             data = _utils.pin_memory.pin_memory(data)

[/usr/local/lib/python3.7/dist-packages/torch/utils/data/_utils/fetch.py](https://localhost:8080/#) in fetch(self, possibly_batched_index)
     50         else:
     51             data = self.dataset[possibly_batched_index]
---> 52         return self.collate_fn(data)

[/usr/local/lib/python3.7/dist-packages/torch_geometric/loader/dataloader.py](https://localhost:8080/#) in __call__(self, batch)
     18         if isinstance(elem, BaseData):
     19             return Batch.from_data_list(batch, self.follow_batch,
---> 20                                         self.exclude_keys)
     21         elif isinstance(elem, torch.Tensor):
     22             return default_collate(batch)

[/usr/local/lib/python3.7/dist-packages/torch_geometric/data/batch.py](https://localhost:8080/#) in from_data_list(cls, data_list, follow_batch, exclude_keys)
     72             add_batch=not isinstance(data_list[0], Batch),
     73             follow_batch=follow_batch,
---> 74             exclude_keys=exclude_keys,
     75         )
     76 

[/usr/local/lib/python3.7/dist-packages/torch_geometric/data/collate.py](https://localhost:8080/#) in collate(cls, data_list, increment, add_batch, follow_batch, exclude_keys)
     83             # Collate attributes into a unified representation:
     84             value, slices, incs = _collate(attr, values, data_list, stores,
---> 85                                            increment)
     86 
     87             if isinstance(value, Tensor) and value.is_cuda:

[/usr/local/lib/python3.7/dist-packages/torch_geometric/data/collate.py](https://localhost:8080/#) in _collate(key, values, data_list, stores, increment)
    131         slices = cumsum([value.size(cat_dim or 0) for value in values])
    132         if increment:
--> 133             incs = get_incs(key, values, data_list, stores)
    134             if incs.dim() > 1 or int(incs[-1]) != 0:
    135                 values = [

[/usr/local/lib/python3.7/dist-packages/torch_geometric/data/collate.py](https://localhost:8080/#) in get_incs(key, values, data_list, stores)
    228         repeats = torch.stack(repeats, dim=0)
    229     else:
--> 230         repeats = torch.tensor(repeats)
    231     return cumsum(repeats[:-1])

RuntimeError: Could not infer dtype of NoneType
```

Can you please help me fix it? Thanks in advance!

**Comment by rusty1s** [comment]
Can you share some more information about your data? It looks like some attribute in one of your `data` objects is `None`, while others hold PyTorch tensors.

**Comment by wzm2256** [comment]
It is a pitty to see that this issue has been unsolved for two years... like a lot of other issues..

The maintainers answer is not correct, because all data is already shown in this post and none of them is None.

#### Why it happens?
This issue is likely to be caused by the hidden behavior in the `__inc__` function https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/data/data.py#L656
 Simply said, when `'index' in key or key == 'face'`, the increased number is `self.num_nodes`, which is `None` in your case, because you didn't set it explicitly, and it can not be inferred from your data (There is no edge_index attribute in your data). Then this None will be converted to Tensor, which causes the error you saw.

#### How to solve it?
My suggestion is to explicitly set `num_nodes attribute` as suggested in the following annotation in the same source file if the default behavior is what you want. Otherwise I suggest modifying the `__inc__` function as suggested in https://pytorch-geometric.readthedocs.io/en/stable/advanced/batching.html where the description is quite vague, and you need to understand the input/output of the `__inc__` file in the source code.


"            The number of nodes in the data object is automatically inferred
            in case node-level attributes are present, *e.g.*, :obj:`data.x`.
            In some cases, however, a graph may only be given without any
            node-level attributes.
            :pyg:`PyG` then *guesses* the number of nodes according to
            :obj:`edge_index.max().item() + 1`.
            However, in case there exists isolated nodes, this number does not
            have to be correct which can result in unexpected behavior.
            Thus, we recommend to set the number of nodes in your data object
            explicitly via :obj:`data.num_nodes = ...`.
            You will be given a warning that requests you to do so."

**Comment by Wapiti08** [comment]
I had the same error, but the error is still there even I explicitly set num_nodes

**Reply by Wapiti08** [reply]
> I had the same error, but the error is still there even I explicitly set num_nodes

I got it fixed. The problem is on my data structure, not the library itself.
